src/write_resume_data.cpp

```diff
@@ -154,9 +109,161 @@ namespace libtorrent {
 							ret_mask.push_back(bit ? '1' : '0');
 					}
 				}
+
+				if (!(flags & write_resume_only) && atp.ti)
+				{
+					file_storage const& fs = atp.ti->files();
+					if (!fs.pad_file_at(f) && fs.file_size(f) > fs.piece_length())
+					{
+						aux::merkle_tree t(fs.file_num_blocks(f)
+							, fs.piece_length() / default_block_size, fs.root_ptr(f));
+
+						if (f < atp.merkle_tree_mask.end_index() && !atp.merkle_tree_mask[f].empty())
+						{
+							t.load_sparse_tree(tree, atp.merkle_tree_mask[f]);
+						}
+						else
+						{
+							t.load_tree(tree);
+						}
+
+						auto const piece_layer = t.get_piece_layer();
+						auto& layer = piece_layers[t.root().to_string()].string();
+
+						for (auto const& h : piece_layer)
+						{
+							layer += h.to_string();
+						}
+					}
+				}
+			}
+
+			if (!piece_layers.empty())
+			{
+				ret["piece layers"] = piece_layers;
```

Comment:
Can you confirm that if `piece_layers` is not empty at this point, then it contains complete/valid "piece layers" data?

Reply:
it may not be complete. There may be files missing piece layers, but the piece layers that are there are correct.

I suppose that's a problem since the loading of a torrent requires all files to be there if it's present

Reply:
I mean you should write "piece layers" field only in the case when it is fully valid (i.e. it produces valid torrent file so it can be correctly loaded later).

Reply:
I was thinking of relaxing the loading function instead, to accept some files not having the piece layer

Reply:
see most recent commit to https://github.com/arvidn/libtorrent/pull/6326

Reply:
>I was thinking of relaxing the loading function instead, to accept some files not having the piece layer

You didn't understand me.
You should either produce fully valid "piece layers" field or produce no such field either.
IIRC, "loading function" still fails if there is invalid "piece layers" that is correct behavior.

Reply:
the benefit of allowing incomplete piece layers when loading torrents is that some hashes are better than none, and not producing a piece layer here means that if any file is "deselected", it won't ever save any hashes when creating a .torrent file.

I think it's reasonable to accept that some files don't have hashes when loading a torrent, I don't see a strong reason to implement "all or nothing" behavior.

Reply:
1. There is a "valid torrent file" requirements in BitTorrent v2 protocol specifications so the function that is intended to create real torrent files should follow it (i.e. create v2/hybrid torrent only with fully valid "piece layers") so that any compliant client can load it.
2. There is nothing wrong to allow load torrent files w/o "piece layers" since libtorrent can process it anyway.
3. We need to produce "incomplete" torrent files only for internal use (to split resume data). Since resume data stores merkle tree data anyway we don't need to store "piece layers" separately, isn't it? So if you provide separate function to produce such "internal purpose" torrent files you shouldn't either care to store "piece layers" there. I wouldn't even call it `write_torrent_file` to eliminate confusion.

Reply:
> There is a "valid torrent file" requirements in BitTorrent v2 protocol specifications so the function that is intended to create real torrent files should follow it (i.e. create v2/hybrid torrent only with fully valid "piece layers") so that any compliant client can load it.

That's reasonable. It's a larger issue than producing incomplete piece layers though. Omitting the piece layer would be equally invalid. This function would need a failure path, either through exception or an `error_code`

Reply:
So how about par.3 above?

Reply:
right, if you were to just include the info-dict in the resume data, you wouldn't need to create .torrent files at all. (i.e. specify that flag to `save_resume_data()`)

Reply:
>if you were to just include the info-dict in the resume data, you wouldn't need to create .torrent files at all.

We found it irrational to overwrite immutable information (info-dict) every time we save resume data.

Personally, I recognize this problem even deeper.
We are dealing with two classes of information about the torrent that we want to store:
1. Immutable or rarely changed (usually changed by event/action) information, e.g. info-dict, save path, file names, etc.
2. Frequently/continuously changed information (usually regardless of events or actions on the user side) information, e.g. data of downloaded pieces, time counters (active, downloading, seeding, etc.).

It is obvious (at least to me) that:
1. It would be more convenient to store these two classes of information separately;
2. It would be more convenient to receive information of different classes in different ways:
2.1 Each item of information of the first class should be available both on demand and in a notifying manner (i.e., when an event occurs, as a result of which the information changes, libtorrent should send a notification),
2.2 Information of the second class should be available on demand, so that the user side can determine the desired frequency of updates.